vnext/ReactWindowsCore/Modules/AsyncStorageModuleWin32.cpp

```diff
@@ -187,154 +255,215 @@ std::map<std::string, dynamic> AsyncStorageModuleWin32::getConstants() {
           Method("multiSet", this, &AsyncStorageModuleWin32::multiSet),
           Method("multiRemove", this, &AsyncStorageModuleWin32::multiRemove),
           Method("clear", this, &AsyncStorageModuleWin32::clear),
           Method("getAllKeys", this, &AsyncStorageModuleWin32::getAllKeys)};
 }
 
-void AsyncStorageModuleWin32::multiGet(folly::dynamic params, Callback callback) {
-  folly::dynamic result = folly::dynamic::array;
-  auto &&args = params[0];
-  if (args.size() == 0) {
-    callback({{}, result});
+void AsyncStorageModuleWin32::multiGet(folly::dynamic args, Callback jsCallback) {
+  auto &keys = args[0];
+  if (keys.size() == 0) {
+    jsCallback({{}, {}});
     return;
   }
-  if (!CheckSize(m_db, args, callback)) {
+  AddTask(DBTask::Type::multiGet, std::move(keys), std::move(jsCallback));
+}
+void AsyncStorageModuleWin32::multiSet(folly::dynamic args, Callback jsCallback) {
+  auto &kvps = args[0];
+  if (kvps.size() == 0) {
+    jsCallback({});
     return;
   }
-  auto argCount = static_cast<int>(args.size());
+  AddTask(DBTask::Type::multiSet, std::move(kvps), std::move(jsCallback));
+}
+void AsyncStorageModuleWin32::multiRemove(folly::dynamic args, Callback jsCallback) {
+  auto &keys = args[0];
+  if (keys.size() == 0) {
+    jsCallback({{}, {}});
+    return;
+  }
+  AddTask(DBTask::Type::multiRemove, std::move(keys), std::move(jsCallback));
+}
+void AsyncStorageModuleWin32::clear(folly::dynamic, Callback jsCallback) {
+  AddTask(DBTask::Type::clear, std::move(jsCallback));
+}
+void AsyncStorageModuleWin32::getAllKeys(folly::dynamic, Callback jsCallback) {
+  AddTask(DBTask::Type::getAllKeys, std::move(jsCallback));
+}
+
+// Under the lock, add a task to m_tasks and, if no async task is in progress,
+// schedule it
+void AsyncStorageModuleWin32::AddTask(
+    AsyncStorageModuleWin32::DBTask::Type type,
+    folly::dynamic &&args,
+    Callback &&jsCallback) {
+  winrt::slim_lock_guard guard(m_lock);
+  m_tasks.emplace_back(type, std::move(args), std::move(jsCallback));
+  if (!m_action)
+    m_action = RunTasks();
+}
+
+// On a background thread, while the async task  has not been cancelled and
+// there are more tasks to do, run the tasks. When there are either no more
+// tasks or cancellation has been requested, set m_action to null to report
+// that and complete the coroutine. N.B., it is important that detecting that
+// m_tasks is empty and acknowledging completion is done atomically; otherwise
+// there would be a race between the background task detecting m_tasks.empty()
+// and AddTask checking the coroutine is running.
+winrt::Windows::Foundation::IAsyncAction AsyncStorageModuleWin32::RunTasks() {
+  auto cancellationToken = co_await winrt::get_cancellation_token();
+  co_await winrt::resume_background();
+  while (!cancellationToken()) {
+    decltype(m_tasks) tasks;
+    sqlite3 *db{nullptr};
+    {
+      winrt::slim_lock_guard guard(m_lock);
+      if (m_tasks.empty()) {
```

Comment:
I think we have a lifetime issue here. I.e. to my understanding this will use the raw `this` pointer which may not be valid after resuming on another thread.

Reply:
Oh nevermind. I missed that we waited on destruction for this to finish.

Reply:
No, actually, it's safe, because the destructor blocks if `m_action` is non-null and waits for the coroutine to acknowledge cancellation. And `m_action` will be non-null when `AddTask` releases `m_lock` (when `guard` goes out of scope). I 100% concede that this is insanely subtle; however, it appears that modules are stored as `unique_ptr`s in the Instance, so I have no idea how else to do it. I'm not sure what the "official" story is for writing thread-safe multi-threaded modules.